File: src/main/stats.js
// Per-game stats (playtime, last launch, install size) + integrity checking.
// Persisted as stats.json in the launcher's userData dir.
'use strict';

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { app } = require('electron');

const statsPath = () => path.join(app.getPath('userData'), 'stats.json');
const integrityPath = () => path.join(app.getPath('userData'), 'integrity.json');

function readJson(file, fallback) {
  try {
    return JSON.parse(fs.readFileSync(file, 'utf8'));
  } catch {
    return fallback;
  }
}
function writeJson(file, data) {
  fs.mkdirSync(path.dirname(file), { recursive: true });
  fs.writeFileSync(file, JSON.stringify(data, null, 2));
}

function loadStats() {
  return readJson(statsPath(), {});
}
function saveStats(all) {
  writeJson(statsPath(), all);
}

// Get a game's stats block, creating a default.
function gameStats(gameId) {
  const all = loadStats();
  return all[gameId] || { playSeconds: 0, lastPlayed: 0, sizeBytes: 0 };
}

// Add elapsed seconds to a game's total playtime and stamp the launch time.
function addPlaytime(gameId, seconds, launchedAt) {
  const all = loadStats();
  const s = all[gameId] || { playSeconds: 0, lastPlayed: 0, sizeBytes: 0 };
  s.playSeconds += Math.max(0, Math.round(seconds));
  s.lastPlayed = launchedAt || Date.now();
  all[gameId] = s;
  saveStats(all);
}

function setLastPlayed(gameId, when) {
  const all = loadStats();
  const s = all[gameId] || { playSeconds: 0, lastPlayed: 0, sizeBytes: 0 };
  s.lastPlayed = when;
  all[gameId] = s;
  saveStats(all);
}

// Recursively total the byte size of a directory.
function dirSize(dir) {
  let total = 0;
  let stack = [dir];
  while (stack.length) {
    const cur = stack.pop();
    let entries;
    try {
      entries = fs.readdirSync(cur, { withFileTypes: true });
    } catch {
      continue;
    }
    for (const e of entries) {
      const p = path.join(cur, e.name);
      if (e.isDirectory()) stack.push(p);
      else {
        try {
          total += fs.statSync(p).size;
        } catch {
          /* ignore */
        }
      }
    }
  }
  return total;
}

function updateSize(gameId, dir) {
  const all = loadStats();
  const s = all[gameId] || { playSeconds: 0, lastPlayed: 0, sizeBytes: 0 };
  s.sizeBytes = dirSize(dir);
  all[gameId] = s;
  saveStats(all);
  return s.sizeBytes;
}

// ---- Integrity -----------------------------------------------------------
// Snapshot sha256 of key files (the exe + every mod). Compared on launch so a
// tampered install is detected. We snapshot right after install/update.
function keyFiles(gameDir, exeName) {
  const files = [path.join(gameDir, exeName)];
  const modsDir = path.join(gameDir, 'geode', 'mods');
  try {
    for (const f of fs.readdirSync(modsDir)) {
      if (f.endsWith('.geode')) files.push(path.join(modsDir, f));
    }
  } catch {
    /* no mods dir */
  }
  return files;
}

function hashFile(file) {
  const buf = fs.readFileSync(file);
  return crypto.createHash('sha256').update(buf).digest('hex');
}

function snapshotIntegrity(gameId, gameDir, exeName) {
  const map = {};
  for (const f of keyFiles(gameDir, exeName)) {
    try {
      map[path.relative(gameDir, f)] = hashFile(f);
    } catch {
      /* ignore missing */
    }
  }
  const all = readJson(integrityPath(), {});
  all[gameId] = map;
  writeJson(integrityPath(), all);
  return map;
}

// Returns { ok, changed: [relPaths] }. ok=true when nothing was tampered with.
function verifyIntegrity(gameId, gameDir, exeName) {
  const all = readJson(integrityPath(), {});
  const snap = all[gameId];
  if (!snap) return { ok: true, changed: [], noSnapshot: true };
  const changed = [];
  for (const [rel, expected] of Object.entries(snap)) {
    const abs = path.join(gameDir, rel);
    let actual = null;
    try {
      actual = hashFile(abs);
    } catch {
      actual = null;
    }
    if (actual !== expected) changed.push(rel);
  }
  return { ok: changed.length === 0, changed };
}

module.exports = {
  gameStats,
  addPlaytime,
  setLastPlayed,
  updateSize,
  snapshotIntegrity,
  verifyIntegrity,
};

File: src/Stats.cs
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace PetusLauncher;

record GameStat(long PlaySeconds, long LastPlayed, long SizeBytes);

record IntegrityResult(bool Ok, List<string> Changed, bool NoSnapshot);

static class Stats
{
    static JsonObject Read(string file)
    {
        try { return JsonNode.Parse(File.ReadAllText(file)) as JsonObject ?? new JsonObject(); }
        catch { return new JsonObject(); }
    }
    static void Write(string file, JsonObject obj)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(file)!);
        File.WriteAllText(file, obj.ToJsonString(new JsonSerializerOptions { WriteIndented = true }));
    }

    public static GameStat Get(string gameId)
    {
        var all = Read(Config.StatsFile);
        if (all[gameId] is JsonObject s)
            return new GameStat(
                (long?)s["playSeconds"] ?? 0,
                (long?)s["lastPlayed"] ?? 0,
                (long?)s["sizeBytes"] ?? 0);
        return new GameStat(0, 0, 0);
    }

    static void Set(string gameId, GameStat st)
    {
        var all = Read(Config.StatsFile);
        all[gameId] = new JsonObject
        {
            ["playSeconds"] = st.PlaySeconds,
            ["lastPlayed"] = st.LastPlayed,
            ["sizeBytes"] = st.SizeBytes,
        };
        Write(Config.StatsFile, all);
    }

    public static void AddPlaytime(string gameId, double seconds, long launchedAt)
    {
        var s = Get(gameId);
        Set(gameId, s with { PlaySeconds = s.PlaySeconds + (long)Math.Max(0, Math.Round(seconds)), LastPlayed = launchedAt });
    }

    public static void SetLastPlayed(string gameId, long when)
    {
        var s = Get(gameId);
        Set(gameId, s with { LastPlayed = when });
    }

    public static long UpdateSize(string gameId, string dir)
    {
        long size = DirSize(dir);
        var s = Get(gameId);
        Set(gameId, s with { SizeBytes = size });
        return size;
    }

    static long DirSize(string dir)
    {
        long total = 0;
        if (!Directory.Exists(dir)) return 0;
        try
        {
            foreach (var f in Directory.EnumerateFiles(dir, "*", SearchOption.AllDirectories))
            {
                try { total += new FileInfo(f).Length; } catch { }
            }
        }
        catch { }
        return total;
    }

    // ---- integrity: sha256 of exe + every .geode mod ----
    static IEnumerable<string> KeyFiles(string gameDir)
    {
        yield return Path.Combine(gameDir, Config.ExeName);
        var mods = Path.Combine(gameDir, "geode", "mods");
        if (Directory.Exists(mods))
            foreach (var f in Directory.EnumerateFiles(mods, "*.geode"))
                yield return f;
    }

    static string Hash(string file)
    {
        using var sha = SHA256.Create();
        using var fs = File.OpenRead(file);
        return Convert.ToHexString(sha.ComputeHash(fs));
    }

    public static void Snapshot(string gameId, string gameDir)
    {
        var map = new JsonObject();
        foreach (var f in KeyFiles(gameDir))
        {
            try { map[Path.GetRelativePath(gameDir, f)] = Hash(f); } catch { }
        }
        var all = Read(Config.IntegrityFile);
        all[gameId] = map;
        Write(Config.IntegrityFile, all);
    }

    public static IntegrityResult Verify(string gameId, string gameDir)
    {
        return VerifyWithProgress(gameId, gameDir, null);
    }

    // Same as Verify but calls `progress(done, total)` as it hashes each file so
    // the UI can drive a progress bar.
    public static IntegrityResult VerifyWithProgress(string gameId, string gameDir, Action<int, int>? progress)
    {
        var all = Read(Config.IntegrityFile);
        if (all[gameId] is not JsonObject snap)
            return new IntegrityResult(true, new(), true);

        var changed = new List<string>();
        int total = snap.Count, done = 0;
        progress?.Invoke(0, total);
        foreach (var kv in snap)
        {
            var rel = kv.Key;
            var expected = kv.Value?.GetValue<string>();
            var abs = Path.Combine(gameDir, rel);
            string? actual = null;
            try { actual = Hash(abs); } catch { }
            if (actual != expected) changed.Add(rel);
            done++;
            progress?.Invoke(done, total);
        }
        return new IntegrityResult(changed.Count == 0, changed, false);
    }
}
